fem/fe_ray_intersect.cpp

```diff
@@ -0,0 +1,453 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "fe_ray_intersect.hpp"
+
+#include "../mesh/mesh.hpp"     // for mfem::Mesh
+
+#include "../fem/geom.hpp"      // for Geometry
+#include "../fem/intrules.hpp"  // for IntegrationPoint & IntegrationRule
+#include "../fem/fe.hpp"        // for FiniteElement
+#include "../fem/fespace.hpp"   // for FiniteElementSpace
+#include "../fem/eltrans.hpp"   // for ElementTransformation
+
+namespace mfem
+{
+
+static struct params_t
+{
+   int     maxNewtonIterations     = 16;
+   int     maxLineSearchIterations = 5;
+   double  c1  = 10.e-4;
+   double  c2  = 0.9;
+   double  tol = 1.e-12;
+} solver_parameters ;
+
+//------------------------------------------------------------------------------
+//                   INTERNAL HELPER METHODS
+//------------------------------------------------------------------------------
+namespace
+{
+
+template < typename T >
+inline T clamp_value( T val, T min, T max )
+{
+   return ( (val < min) ? min : (val > max) ? max : val );
+}
+
+//------------------------------------------------------------------------------
+template < typename T >
+inline T clamp_min( T val, T min )
+{
+   return ( (val < min) ? min : val );
+}
+
+//------------------------------------------------------------------------------
+void get_physical_nodes( const FiniteElement* fe,
+                         ElementTransformation* T,
+                         DenseMatrix& phys_nodes )
+{
+   MFEM_ASSERT( fe != nullptr, "supplied finite element is null" );
+   MFEM_ASSERT( T != nullptr, "supplied element transformation is null" );
+
+   const IntegrationRule& ref_nodes = fe->GetNodes();
+   const int ndofs = ref_nodes.Size();
+   const int ndims = T->GetSpaceDim();
+   phys_nodes.SetSize( ndims, ndofs );
+   T->Transform( ref_nodes, phys_nodes );
+}
+
+//------------------------------------------------------------------------------
+bool check_convergence( const double* dx, int ndims, double TOL )
+{
+   double l1norm = 0.0;
```

Comment:
Lots of free functions in this file. Is there a natural way to put them in a class, e.g. something like `InverseElementTransformation`?

Reply:
Most of them are private to this implementation and are encapsulated within an anonymous namespace. The methods that are exposed to the application are defined in the ``*.hpp`` file.

They can be put in the class, but, I think from the user's perspective, it would be simpler if they were static methods of the class (see my previous response which relates to this).